app/lib/actions/remote_execution/run_proxy_command.rb

```diff
@@ -5,17 +5,60 @@ class RunProxyCommand < Actions::ProxyAction
       include ::Dynflow::Action::Cancellable
 
       def plan(proxy, hostname, script, options = {})
-        options = { :effective_user => nil }.merge(options)
+        options = default_connection_options
+                  .merge({ :effective_user => nil })
+                  .merge(options)
         super(proxy, options.merge(:hostname => hostname, :script => script))
       end
 
+      def run(event = nil)
+        with_connection_error_handling(event) { super(event) }
+      end
+
       def proxy_action_name
         'Proxy::RemoteExecution::Ssh::CommandAction'
       end
 
       def on_data(data)
-        super(data)
-        error! _("Script execution failed") if failed_run?
+        if data[:result] == 'initialization_error'
+          handle_connection_exception(data[:metadata][:exception_class]
+                                       .constantize
+                                       .new(data[:metadata][:exception_message]))
+        else
+          super(data)
+          error! _("Script execution failed") if failed_run?
+        end
+      end
+
+      def trigger_proxy_task
+        suspend do |suspended_action|
+          output[:metadata] ||= {}
+          unless metadata[:timeout]
+            time = Time.now + input[:connection_options][:timeout]
+            @world.clock.ping suspended_action,
+                              time,
+                              Timeout.new
+            metadata[:timeout] = time.to_s
+          end
+          super
+        end
+      end
+
+      def check_task_status
+        if output[:proxy_task_id]
```

Comment:
This logic could also go into https://github.com/theforeman/foreman-tasks/pull/132/files#diff-03dde90ec7e9caa322015dee15f4e956R51